sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala

```diff
@@ -167,7 +167,28 @@ class EquivalentExpressions {
    * Returns all the equivalent sets of expressions.
    */
   def getAllEquivalentExprs: Seq[Seq[Expression]] = {
-    equivalenceMap.values.map(_.toSeq).toSeq
+    equivalenceMap.values.map(_.toSeq).toSeq.sortBy(_.head)(new ExpressionContainmentOrdering)
+  }
+
+  /**
+   * Orders [Expression] by parent/child relations. The child expression is smaller
+   * than parent expression. If there is child-parent relationships among the subexpressions,
+   * we want the child expressions come first than parent expressions, so we can replace
+   * child expressions in parent expressions with subexpression evaluation. Note that
+   * this is not for general expression ordering. For example, two irrelevant expressions
+   * will be considered as e1 < e2 and e2 < e1 by this ordering. But for the usage here,
+   * the order of irrelevant expressions does not matter.
+   */
+  class ExpressionContainmentOrdering extends Ordering[Expression] {
+    override def compare(x: Expression, y: Expression): Int = {
+      if (x.semanticEquals(y)) {
+        0
+      } else if (x.find(_.semanticEquals(y)).isDefined) {
```

Comment:
can we run `TPCDSQuerySuite` and see the time of the query compilation phase? This looks like a very expensive sort.

Reply:
Ok. Let me compare before/after this PR.

Reply:
BTW, I think better approach is to sort after filter (e.g. size > 1 in most use-case), because the number of sub-exprs should be smaller.

Reply:
I changed the call usage of `getAllEquivalentExprs`. So we filter it first and then do sorting.

Reply:
Ran `TPCDSQuerySuite`.

Before (master):
```
23.233160578 seconds 
22.501728011 seconds
23.547332524 seconds
```

After:

```
23.995751468 seconds 
22.262832936 seconds
21.503776059 seconds  
```

I don't see significant difference there.